pkg/skaffold/kubernetes/sync.go

```diff
@@ -34,60 +34,76 @@ import (
+		}
 	}
-	if err := perform(s.Image, s.Delete, deleteFileFn); err != nil {
-		return errors.Wrap(err, "deleting files")
+
+	if len(s.Delete) > 0 {
+		logrus.Infoln("Deleting files:", s.Delete, "from", s.Image)
+
+		if err := perform(ctx, s.Image, s.Delete, deleteFileFn); err != nil {
+			return errors.Wrap(err, "deleting files")
+		}
 	}
-	return nil
-}
 
-func deleteFileFn(pod v1.Pod, container v1.Container, src, dst string) *exec.Cmd {
-	return exec.Command("kubectl", "exec", pod.Name, "--namespace", pod.Namespace, "-c", container.Name, "--", "rm", "-rf", dst)
+	return nil
 }
 
-func copyFileFn(pod v1.Pod, container v1.Container, src, dst string) *exec.Cmd {
-	return exec.Command("kubectl", "cp", src, fmt.Sprintf("%s/%s:%s", pod.Namespace, pod.Name, dst), "-c", container.Name)
+func deleteFileFn(ctx context.Context, pod v1.Pod, container v1.Container, src, dst string) *exec.Cmd {
+	return exec.CommandContext(ctx, "kubectl", "exec", pod.Name, "--namespace", pod.Namespace, "-c", container.Name, "--", "rm", "-rf", dst)
 }
 
-func labelSelector() string {
-	var reqs []string
-	for k, v := range constants.Labels.DefaultLabels {
-		reqs = append(reqs, fmt.Sprintf("%s=%s", k, v))
-	}
-	return strings.Join(reqs, ",")
+func copyFileFn(ctx context.Context, pod v1.Pod, container v1.Container, src, dst string) *exec.Cmd {
+	return exec.CommandContext(ctx, "kubectl", "cp", src, fmt.Sprintf("%s/%s:%s", pod.Namespace, pod.Name, dst), "-c", container.Name)
 }
 
-func perform(image string, files map[string]string, cmdFn func(v1.Pod, v1.Container, string, string) *exec.Cmd) error {
-	logrus.Info("Syncing files:", files)
+func perform(ctx context.Context, image string, files map[string]string, cmdFn func(context.Context, v1.Pod, v1.Container, string, string) *exec.Cmd) error {
 	client, err := Client()
 	if err != nil {
 		return errors.Wrap(err, "getting k8s client")
 	}
-	pods, err := client.CoreV1().Pods("").List(meta_v1.ListOptions{
-		LabelSelector: labelSelector(),
-	})
+
+	pods, err := client.CoreV1().Pods("").List(meta_v1.ListOptions{})
 	if err != nil {
 		return errors.Wrap(err, "getting pods")
 	}
+
+	var performed int32
+	var e errgroup.Group
+
 	for _, p := range pods.Items {
 		for _, c := range p.Spec.Containers {
-			if c.Image == image {
-				var e errgroup.Group
-				for src, dst := range files {
-					src, dst := src, dst
-					e.Go(func() error {
-						cmd := cmdFn(p, c, src, dst)
-						return util.RunCmd(cmd)
-					})
-				}
-				if err := e.Wait(); err != nil {
-					return errors.Wrap(err, "syncing files")
-				}
+			if c.Image != image {
+				continue
+			}
+
+			for src, dst := range files {
+				cmd := cmdFn(ctx, p, c, src, dst)
+
+				e.Go(func() error {
+					if err := util.RunCmd(cmd); err != nil {
+						return err
+					}
+
+					atomic.AddInt32(&performed, 1)
+					return nil
+				})
 			}
 		}
 	}
+
+	if err := e.Wait(); err != nil {
+		return errors.Wrap(err, "syncing files")
+	}
+
+	if int(performed) != len(files) {
```

Comment:
I don't think you need to count here. errgroup handles this

Reply:
Hm, I guess this is a bit tricky. What if we have 2 replicas?

Reply:
Good point, I'll see how to fix it. Each file should be synced at least once, I guess.

Reply:
Yeah I think thats the best we can do. But making sure we sync at least once is a good check.